### PR TITLE
Return right BMC state to the redfish client

### PR DESCRIPTION
**Pull request by raviteja-b**
Currently BMC PowerState hardcoded to "On" on redfish interface.

this commit adds check to display bmc state based on current bmc state on state manager D-bus 

This is to fix HMC-BMC defect SW541358

Tested by:
GET /redfish/v1/Managers/bmc for different bmc states
Redfish validator passed

### CODE REVIEW

File: redfish-core/lib/managers.hpp [hunk 1]
```diff
@@ -1922,6 +1922,61 @@ inline void setDateTime(std::shared_ptr<bmcweb::AsyncResp> aResp,
...
+ */
+inline void getBMCState(const std::shared_ptr<bmcweb::AsyncResp>& aResp)
+{
+    BMCWEB_LOG_DEBUG << "Get BMC information.";
+    crow::connections::systemBus->async_method_call(
+        [aResp](const boost::system::error_code ec,
+                const std::variant<std::string>& bmcState) {
+            if (ec)
+            {
+                BMCWEB_LOG_DEBUG << "DBUS response error " << ec;
```
Comment (sunharis) [inline redfish-core/lib/managers.hpp:1940]: Change this to error.

File: redfish-core/lib/managers.hpp [hunk 2]
```diff
@@ -1922,6 +1922,61 @@ inline void setDateTime(std::shared_ptr<bmcweb::AsyncResp> aResp,
...
+                {
+                    aResp->res.jsonValue["PowerState"] = "Off";
+                    aResp->res.jsonValue["Status"]["State"] = "Disabled";
+                }
+                else
+                {
+                    aResp->res.jsonValue["PowerState"] = "Off";
+                    aResp->res.jsonValue["Status"]["State"] = "Disabled";
+                }
+                aResp->res.jsonValue["Status"]["Health"] = "OK";
```
Comment (sunharis) [inline redfish-core/lib/managers.hpp:1972]: Probably we need to say health is not good if the state is Quiesced ?

**Review by sunharis — CHANGES_REQUESTED** [review]

**Comment by sunharis** [comment]
CI should pass

File: redfish-core/lib/managers.hpp (hunk 2 above)
Comment (raviteja-b) [reply redfish-core/lib/managers.hpp:1972]: You mean Health ="Critical"?

**Comment by raviteja-b** [comment]
jenkins run tests please

File: redfish-core/lib/managers.hpp (hunk 1 above)
Comment (raviteja-b) [reply redfish-core/lib/managers.hpp:1940]: ok

File: redfish-core/lib/managers.hpp (hunk 2 above)
Comment (raviteja-b) [reply redfish-core/lib/managers.hpp:1972]: @gtmills what do you suggest?

**Comment by geissonator** [comment]
jenkins run tests please

File: redfish-core/lib/managers.hpp [hunk 3]
```diff
@@ -1922,6 +1922,61 @@ inline void setDateTime(std::shared_ptr<bmcweb::AsyncResp> aResp,
...
+        [aResp](const boost::system::error_code ec,
+                const std::variant<std::string>& bmcState) {
+            if (ec)
+            {
+                BMCWEB_LOG_DEBUG << "DBUS response error " << ec;
+                messages::internalError(aResp->res);
+                return;
+            }
+
+            const std::string* s = std::get_if<std::string>(&bmcState);
```
Comment (gtmills) [inline redfish-core/lib/managers.hpp:1945]: please don't use a single var name like s. how about state ?

File: redfish-core/lib/managers.hpp [hunk 4]
```diff
@@ -1922,6 +1922,61 @@ inline void setDateTime(std::shared_ptr<bmcweb::AsyncResp> aResp,
...
+            BMCWEB_LOG_DEBUG << "BMC state: " << *s;
+            if (s != nullptr)
```
Comment (gtmills) [inline redfish-core/lib/managers.hpp:1947]: These are typically reverse checks in bmcweb 

if (state == nullptr)
{
internalError; 
return;

}

File: redfish-core/lib/managers.hpp (hunk 2 above)
Comment (gtmills) [reply redfish-core/lib/managers.hpp:1972]: You have an if and else why do you need this line ?

File: redfish-core/lib/managers.hpp [hunk 5]
```diff
@@ -1922,6 +1922,61 @@ inline void setDateTime(std::shared_ptr<bmcweb::AsyncResp> aResp,
...
+                               "Quiesced")
+                {
+                    aResp->res.jsonValue["PowerState"] = "On";
+                    aResp->res.jsonValue["Status"]["State"] = "Quiesced";
+                }
+                else if (*s == "xyz.openbmc_project.State.BMC.BMCState."
+                               "NotReady")
+                {
+                    aResp->res.jsonValue["PowerState"] = "Off";
+                    aResp->res.jsonValue["Status"]["State"] = "Disabled";
```
Comment (gtmills) [inline redfish-core/lib/managers.hpp:1965]: Does starting make more sense here? 
                "Starting": "This function or resource is starting.",

**Comment by gtmills** [comment]
Validator passes ? 
Can you push this upstream?

**Comment by gtmills** [comment]
jenkins run test please

File: redfish-core/lib/managers.hpp (hunk 4 above)
Comment (raviteja-b) [reply redfish-core/lib/managers.hpp:1947]: done

File: redfish-core/lib/managers.hpp (hunk 2 above)
Comment (raviteja-b) [reply redfish-core/lib/managers.hpp:1981]: done

File: redfish-core/lib/managers.hpp (hunk 1 above)
Comment (raviteja-b) [reply redfish-core/lib/managers.hpp]: done

File: redfish-core/lib/managers.hpp (hunk 3 above)
Comment (raviteja-b) [reply redfish-core/lib/managers.hpp]: done

File: redfish-core/lib/managers.hpp (hunk 5 above)
Comment (raviteja-b) [reply redfish-core/lib/managers.hpp]: done

**Review by sunharis — APPROVED** [review]
Please update the upstream commit link here for reference

**Comment by raviteja-b** [comment]
Upstream commit https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/50318

**Review by asmithakarun — APPROVED** [review]

File: redfish-core/lib/managers.hpp [hunk 6]
```diff
@@ -1922,6 +1922,70 @@ inline void setDateTime(std::shared_ptr<bmcweb::AsyncResp> aResp,
...
+                               "Quiesced")
+            {
+                aResp->res.jsonValue["PowerState"] = "On";
+                aResp->res.jsonValue["Status"]["State"] = "Quiesced";
+                aResp->res.jsonValue["Status"]["Health"] = "Critical";
+            }
+            else if (*state == "xyz.openbmc_project.State.BMC.BMCState."
+                               "NotReady")
+            {
+                aResp->res.jsonValue["PowerState"] = "Starting";
```
Comment (gtmills) [inline redfish-core/lib/managers.hpp:1966]: This is not a valid PowerState. PowerState shouldprobably be PoweringOn.
```
        "PowerState": {
            "enum": [
                "On",
                "Off",
                "PoweringOn",
                "PoweringOff",
                "Paused"
            ],
```            
            State needs to be Starting
            
 ```           
                    "State": {
            "enum": [
                "Enabled",
                "Disabled",
                "StandbyOffline",
                "StandbySpare",
                "InTest",
                "Starting",
                "Absent",
                "UnavailableOffline",
                "Deferring",
                "Quiesced",
                "Updating",
                "Qualified"
            ],
```

Comment (raviteja-b) [reply redfish-core/lib/managers.hpp]: done

File: redfish-core/lib/managers.hpp [hunk 7]
```diff
@@ -1922,6 +1922,70 @@ inline void setDateTime(std::shared_ptr<bmcweb::AsyncResp> aResp,
...
+            else if (*state == "xyz.openbmc_project.State.BMC.BMCState."
+                               "UpdateInProgress")
+            {
+                aResp->res.jsonValue["PowerState"] = "PoweringOn";
+                aResp->res.jsonValue["Status"]["State"] = "Updating";
+                aResp->res.jsonValue["Status"]["Health"] = "OK";
+            }
+            else
+            {
+                aResp->res.jsonValue["PowerState"] = "Off";
```
Comment (gtmills) [inline redfish-core/lib/managers.hpp:1979]: Isn't this else just an error ?

Comment (gtmills) [reply redfish-core/lib/managers.hpp]: I guess that is what we do here: 
https://github.com/ibm-openbmc/bmcweb/blob/master/redfish-core/lib/systems.hpp#L624

**Pull request merged**